error [cli-launch-master] Unexpected error during benchmark execution: Error: Cannot find module 'fabric-ca-client/package

<!--- Provide a general summary of the issue in the Title above -->

## Context
<!--- Provide a more detailed introduction to the issue itself, and why you consider it to be a bug -->
benchmark fabric 2.0.0 with latest stable caliper image 0.3.0
## Expected Behavior
<!--- Tell us what should happen -->
caliper container runs and generates loads into STU
## Actual Behavior
<!--- Tell us what happens instead -->
caliper    | added 545 packages from 191 contributors in 968.321s
caliper    | 2020.03.16-01:51:23.915 +0000 info  [cli-launch-master] Set workspace path: /hyperledger/caliper/workspace
caliper    | 2020.03.16-01:51:23.915 +0000 info  [cli-launch-master] Set benchmark configuration path: /hyperledger/caliper/workspace/benchmark/benchconfig.yaml
caliper    | 2020.03.16-01:51:23.916 +0000 info  [cli-launch-master] Set network configuration path: /hyperledger/caliper/workspace/benchmark/networkconfig.yaml
caliper    | 2020.03.16-01:51:23.916 +0000 info  [cli-launch-master] Set SUT type: fabric:latest
caliper    | 2020.03.16-01:51:24.037 +0000 info  [caliper-engine] Starting benchmark flow
caliper    | 2020.03.16-01:51:24.041 +0000 **error [cli-launch-master] Unexpected error during benchmark execution: Error: Cannot find module 'fabric-ca-client/package'**
caliper    | Usage:
caliper    |  caliper launch master --caliper-bind-sut fabric:1.4.1 [other options]
caliper    |
caliper    | Options:
caliper    |   --help, -h           Show usage information  [boolean]
caliper    |   --version            Show version information  [boolean]
caliper    |   --caliper-bind-sut   The name and version of the platform to bind to  [string]
caliper    |   --caliper-bind-cwd   The working directory for performing the SDK install  [string]
caliper    |   --caliper-bind-args  Additional arguments to pass to "npm install". Use the "=" notation when setting this parameter  [string]
caliper    |   --caliper-bind-file  Yaml file to override default (supported) package versions when binding an SDK  [string]
caliper    |
caliper    | { Error: Cannot find module 'fabric-ca-client/package'
caliper    |     at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
caliper    |     at Function.Module._load (internal/modules/cjs/loader.js:562:25)
caliper    |     at Module.require (internal/modules/cjs/loader.js:692:17)
caliper    |     at require (internal/modules/cjs/helpers.js:25:18)
caliper    |     at new Fabric (/home/node/.npm-global/lib/node_modules/@hyperledger/caliper-cli/node_modules/@hyperledger/caliper-fabric/lib/fabric.js:32:32)
caliper    |     at CaliperEngine.adapterFactory (/home/node/.npm-global/lib/node_modules/@hyperledger/caliper-cli/node_modules/@hyperledger/caliper-fabric/lib/adapterFactory.js:26:19)
caliper    |     at CaliperEngine.run (/home/node/.npm-global/lib/node_modules/@hyperledger/caliper-cli/node_modules/@hyperledger/caliper-core/lib/master/caliper-engine.js:94:34)
caliper    |     at Function.handler (/home/node/.npm-global/lib/node_modules/@hyperledger/caliper-cli/lib/launch/lib/launchMaster.js:62:43)
caliper    |     at process._tickCallback (internal/process/next_tick.js:68:7) code: 'MODULE_NOT_FOUND' }
caliper exited with code 1
## Possible Fix
<!--- Not obligatory, but suggest a fix or reason for the bug -->

## Steps to Reproduce
<!--- Provide a link to a live example, or an unambiguous set of steps to -->
<!--- reproduce this bug include code to reproduce, if relevant -->
1.  docker-compose up



The following is docker-compose.yaml file.

version: '2'

networks:
  caliper_net:
   external:
      name: net_byfn
services:
    caliper:
        container_name: caliper
        image: hyperledger/caliper:0.3.0
        environment:
        - CALIPER_BIND_SUT=fabric:latest
          #- CALIPER_BIND_SDK=1.4.7
        - CALIPER_PROJECTCONFIG=benchmark/caliper.yaml
        - CALIPER_BENCHCONFIG=benchmark/benchconfig.yaml
        - CALIPER_NETWORKCONFIG=benchmark/networkconfig.yaml
        # command for self building image hyperledger/caliper:0.2.0light, containing fabric replated initialization
        command: launch master
        volumes:
        - ./benchmark:/hyperledger/caliper/workspace/benchmark
        - ./invoke.js:/hyperledger/caliper/workspace/invoke.js
        - ../crypto-config:/hyperledger/caliper/workspace/config/crypto-config
        - /etc/hosts:/etc/hosts
        networks:
          - caliper_net

## Existing issues
No

<!-- please include any links to issues here -->

## Context
<!--- How has this bug affected you? What were you trying to accomplish? -->
With caliper image 0.3.0, intend to benchmark fabric 2.0.0.
## Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
* image tag : hyperledger/caliper:0.3.0
* node version:  v12.16.1
* host operating system and version : ubuntu 18.04.3
* docker version： 19.03.5
* docker-compose version：1.17.1

--- a/packages/caliper-core/lib/common/utils/caliper-utils.js
+++ b/packages/caliper-core/lib/common/utils/caliper-utils.js
@@ -148,6 +148,31 @@ class CaliperUtils {
     }
 
     /**
+     * Check if a named module is installed and accessible by caliper
+     * @param {string} moduleName the name of the module to check
+     * @param {Function} requireFunction The "require" function (with appropriate scoping) to use to load the module.
+     * @returns {boolean} boolean value for existence of accessible package
+     */
+    static moduleIsInstalled(moduleName, requireFunction = require) {
+        let modulePath;
+        if (moduleName.startsWith('./') || moduleName.startsWith('/')) {
+            // treat it as an external module, but resolve the path, so it's absolute
+            modulePath = CaliperUtils.resolvePath(moduleName);
+        } else {
+            // treat it as a package dependency (user must install it beforehand)
+            modulePath = moduleName;
+        }
+
+        try {
+            CaliperUtils.loadModule(modulePath, requireFunction);
+            return true;
+        } catch (err) {
+            // If the module can not be found, an error is thrown
+            return false;
+        }
+    }
+
+    /**
      * Utility function to check for singleton values
      * @param {string[]} passedArgs arguments passed by user
      * @param {string[]} uniqueArgs arguments that must be unique

--- a/packages/caliper-fabric/lib/fabric.js
+++ b/packages/caliper-fabric/lib/fabric.js
@@ -28,9 +28,20 @@ const Fabric = class extends BlockchainInterface {
      */
     constructor(workerIndex) {
         super();
-        // Switch adaptors on the fabric-ca-client, which is a common package across all fabric-sdk-node releases
-        const packageVersion = require('fabric-ca-client/package').version;
-        const version = semver.coerce(packageVersion);
+        // Switch adaptors based on installed packages
+        // - will either have fabric-client, or fabric-network
+        let version;
+        if (CaliperUtils.moduleIsInstalled('fabric-client')) {
+            const packageVersion = require('fabric-client/package').version;
+            version = semver.coerce(packageVersion);
+        } else if (CaliperUtils.moduleIsInstalled('fabric-network')) {
+            const packageVersion = require('fabric-network/package').version;
+            version = semver.coerce(packageVersion);
+        } else {
+            const msg = 'Unable to detect required Fabric binding packages';
+            throw new Error(msg);
+        }
+
         const useGateway = ConfigUtil.get(ConfigUtil.keys.Fabric.Gateway.UseGateway, false);
 
         Logger.info(`Initializing ${useGateway ? 'gateway' : 'standard' } adaptor compatible with installed SDK: ${version}`);

